### PR TITLE
rewrite tests and define inplace versions of tovoigt

### PR DESCRIPTION
**Pull request by fredrikekre**
- Remove most `isa` tests and add type assertions instead
- check some more with `@inferred`

### CODE REVIEW

**Comment by fredrikekre** [comment]
On 0.6:
```julia
julia> A = rand(9,9);

julia> @inferred fromvoigt(Tensor{4, 3}, A);
ERROR: return type Tensors.Tensor{4,3,Float64,81} does not match inferred return type Any
Stacktrace:
 [1] error(::String) at ./error.jl:21
```

**Comment by fredrikekre** [comment]
Locally tests take 20 min on 0.5... But Travis test time doesn't seem to be affected...

**Comment by KristofferC** [comment]
With `Pkg.test()`? Perhaps when inlining is off it does less optimizations and thus faster test time?

**Comment by fredrikekre** [comment]
With `Pkg.test()`

**Comment by fredrikekre** [comment]
It seems to be just the testset with construction from Array which is extremely slow though.

**Comment by KristofferC** [comment]
Yeah, I wonder if we can tweak that test somehow.

**Comment by fredrikekre** [comment]
```
 ─────────────────────────────────────────────────────────────────────────────────────
                                              Time                   Allocations      
                                      ──────────────────────   ───────────────────────
           Tot / % measured:               1119s / 93.5%           24.8GiB / 94.9%    

 Section                      ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────────────────────
 tensor ops                        1     487s  46.6%    487s   10.7GiB  45.4%  10.7GiB
   symmetric/skew-symmetric        9     340s  32.5%   37.8s   6.92GiB  29.4%   787MiB
   transpose                       9    84.5s  8.07%   9.39s   1.97GiB  8.37%   224MiB
   outer product                   9    45.3s  4.32%   5.03s   1.33GiB  5.64%   151MiB
   double contraction              9    13.2s  1.26%   1.47s    385MiB  1.59%  42.7MiB
   tovoigt/fromvoigt               9    2.13s  0.20%   237ms   45.0MiB  0.19%  5.00MiB
   dot products                    9    962ms  0.09%   107ms   28.5MiB  0.12%  3.17MiB
   special                         9    627ms  0.06%  69.7ms   13.1MiB  0.05%  1.46MiB
   rotation                        3    467ms  0.04%   156ms   18.5MiB  0.08%  6.16MiB
   cross product                   9   95.2ms  0.01%  10.6ms   1.88MiB  0.01%   214KiB
 constrct Arr                      1     288s  27.5%    288s   5.68GiB  24.1%  5.68GiB
 promotion/conversion              1    66.7s  6.37%   66.7s   1.51GiB  6.42%  1.51GiB
 AD                                1    66.6s  6.36%   66.6s   1.91GiB  8.11%  1.91GiB
   dim 3                           1    57.9s  5.54%   57.9s   1.60GiB  6.79%  1.60GiB
     hessian                       2    34.8s  3.32%   17.4s    842MiB  3.49%   421MiB
     2nd tensor grad               2    7.13s  0.68%   3.57s    245MiB  1.02%   123MiB
     scalar grad                   2    822ms  0.08%   411ms   30.0MiB  0.12%  15.0MiB
   dim 2                           1    5.87s  0.56%   5.87s    228MiB  0.95%   228MiB
     hessian                       2    2.66s  0.25%   1.33s   97.4MiB  0.40%  48.7MiB
     2nd tensor grad               2    1.36s  0.13%   679ms   51.1MiB  0.21%  25.6MiB
     scalar grad                   2    515ms  0.05%   258ms   19.5MiB  0.08%  9.77MiB
   dim 1                           1    2.81s  0.27%   2.81s   90.9MiB  0.38%  90.9MiB
     hessian                       2    1.25s  0.12%   623ms   43.0MiB  0.18%  21.5MiB
     2nd tensor grad               2    633ms  0.06%   316ms   16.2MiB  0.07%  8.10MiB
     scalar grad                   2    358ms  0.03%   179ms   9.13MiB  0.04%  4.56MiB
 indexing                          1    60.0s  5.74%   60.0s   1.25GiB  5.31%  1.25GiB
 tensor identities                 1    32.9s  3.15%   32.9s   1.05GiB  4.46%  1.05GiB
 simple math                       1    16.8s  1.61%   16.8s    571MiB  2.37%   571MiB
 norm, trace, det, inv, eig        1    13.0s  1.24%   13.0s    437MiB  1.81%   437MiB
 constructors                      1    10.6s  1.01%   10.6s    337MiB  1.40%   337MiB
 base vectors                      1    1.95s  0.19%   1.95s   56.3MiB  0.23%  56.3MiB
 constrct func                     1    1.31s  0.13%   1.31s   48.5MiB  0.20%  48.5MiB
 diagm, one                        1    1.28s  0.12%   1.28s   30.8MiB  0.13%  30.8MiB
 exceptions                        1   2.38ms  0.00%  2.38ms   9.52KiB  0.00%  9.52KiB
 ─────────────────────────────────────────────────────────────────────────────────────

```

**Comment by fredrikekre** [comment]
symmetric 340 sec :confused:

**Comment by KristofferC** [comment]
Yeah, I worry less and less about 0.5 though...

**Comment by fredrikekre** [comment]
Well it runs fast on travis so w/e

File: src/utilities.jl [hunk 1]
```diff
@@ -1,7 +1,7 @@
 macro inboundsret(ex)
     return quote
         @inbounds v = $(esc(ex))
-        v
+        return v
```
Comment (fredrikekre) [inline src/utilities.jl]: Thanks for this ;) This only worked if `@inboundsret` was the last expression in the function...

File: src/voigt.jl [hunk 2]
```diff
@@ -91,69 +100,32 @@ julia> fromvoigt(Tensor{2,3}, 1.0:1.0:9.0)
...
-            end
-            I, si = _index_scale(dim, i, j, offdiagscale)
-            J, sj = _index_scale(dim, k, l, offdiagscale)
-            return v[I, J] / (si * sj)
-        end
-    )
+    return TT(function (i, j)
+            i > j && ((i, j) = (j, i))
+            i == j ? (@inboundsret T(v[VOIGT_ORDER[dim][i, j]])) :
+                     (@inboundsret T(v[VOIGT_ORDER[dim][i, j]] / offdiagscale))
```
Comment (fredrikekre) [inline src/voigt.jl]: ... so I didn't understand why this didn't work correctly :)

File: src/utilities.jl (hunk 1 above)
Comment (KristofferC) [inline src/utilities.jl]: XD

Comment (fredrikekre) [inline src/utilities.jl]: more like `@inboundsassign` lul

Comment (KristofferC) [inline src/utilities.jl]: Yes, this was never intended to explicitly return. What I meant was that it "returns" the evaluated expression. I rather have it like it was (but with a rename if you want).

**Comment by fredrikekre** [comment]
Thats why I allowed `AbstractVecOrMat` such that you can send in a view. If we allow offsets we cannot check the input length etc.

File: src/utilities.jl (hunk 1 above)
Comment (fredrikekre) [inline src/utilities.jl]: Hmm... It is only used where `return`ing is the correct thing to do though?

**Comment by fredrikekre** [comment]
Lets just skip input check and use `Base.@propagate_inbounds` instead actually

**Comment by KristofferC** [comment]
Sending a view is slower.

**Comment by KristofferC** [comment]
I wonder why coverage says that `simd.jl` git -20% coverage..

Do you think it is more "julian" to have `offset` start at 1 (which would copy into the index starting at 1). Most stuff in julia seems to start with 1.

**Comment by fredrikekre** [comment]
> I wonder why coverage says that `simd.jl` git -20% coverage..

Idk, `@inferred` screwing with dispatch?

> Do you think it is more "julian" to have `offset` start at 1 (which would copy into the index starting at 1). Most stuff in julia seems to start with 1.

Sure, updated

**Pull request merged**